src/typechecker.erl

```diff
@@ -608,25 +608,36 @@ glb_ty(_Ty1, _Ty2, _A, _TEnv) -> {type(none), constraints:empty()}.
 %% * Replace built-in type synonyms
 %% * Flatten unions and merge overlapping types (e.g. ranges) in unions
 -spec normalize(type(), TEnv :: #tenv{}) -> type().
-normalize({type, _, record, [{atom, _, Name}|Fields]}, TEnv) when length(Fields) > 0 ->
-    NormFields = [type_field_type(FieldName, normalize(Type, TEnv)) 
+normalize(Ty, TEnv) ->
+    normalize_rec(Ty, TEnv, sets:new()).
+
+%% The third argument is a set of user types that we've already unfolded.
+%% It's important that we don't keep unfolding such types because it will
+%% lead to infinite recursion.
+normalize_rec({type, _, record, [{atom, _, Name}|Fields]}, TEnv, Unfolded)
+  when length(Fields) > 0 ->
+    NormFields = [type_field_type(FieldName, normalize_rec(Type, TEnv, Unfolded)) 
         || ?type_field_type(FieldName, Type) <- Fields],
     type_record(Name, NormFields);
-normalize({type, _, union, Tys}, TEnv) ->
-    Types = flatten_unions(Tys, TEnv),
+normalize_rec({type, _, union, Tys}, TEnv, Unfolded) ->
+    Types = flatten_unions(Tys, TEnv, Unfolded),
     case merge_union_types(Types, TEnv) of
         []  -> type(none);
         [T] -> T;
         Ts when length(Ts) > ?union_size_limit -> type(any); % performance hack
         Ts  -> type(union, Ts)
     end;
-normalize({user_type, P, Name, Args} = Type, TEnv) ->
-    case typelib:get_module_from_annotation(P) of
+normalize_rec({user_type, P, Name, Args} = Type, TEnv, Unfolded) ->
+    case sets:is_element({P,Name}, Unfolded) of
```

Comment:
`{P,Name}` may seem a bit strange identity for a type being unfolded. `P` is a (modified) annotation, - which may depend of how Gradualizer uses/stores/handles annotations.

Eg. - running under debugger reveals the this `P` is always `0`, - meaning that all types with the same name will be put into the same "basket".

Probably, not only the name, but also the arity of the type should be considered in this identity - since Erlang allows to have types with the same name by different arities.

An example:

With this diff, this snippet is type-checked:

```erlang
-module(arity_overloading).

-compile([export_all, nowarn_export_all]).

-type pair(A, B) :: {A, B}.
-type pair1(A) :: pair(A, A).

-spec foo(pair1(pair(atom(), atom()))) -> {atom(), atom()}.
foo({X, _}) -> X.
```

But changing `pair1` -> `pair`, makes it ill-typed. (Gradualizer doesn't distinguish between `pair/1` and `pair/2`).

```erlang
-module(arity_overloading).

-compile([export_all, nowarn_export_all]).

-type pair(A, B) :: {A, B}.
-type pair(A) :: pair(A, A).

-spec foo(pair(pair(atom(), atom()))) -> {atom(), atom()}.
foo({X, _}) -> X.
```

```
-module(arity_overloading).

-compile([export_all, nowarn_export_all]).

-type pair(A, B) :: {A, B}.
-type pair(A) :: pair(A, A).

-spec foo(pair(pair(atom(), atom()))) -> {atom(), atom()}.
foo({X, _}) -> X.
```

```
./bin/gradualizer arity_overloading.erl

arity_overloading.erl: The pattern {X, _} on line 9 at column 5 doesn't have the type pair(pair(atom(), atom()))
```

Reply:
Right, shouldn't we store `{Name, Args}` in the set instead?

Reply:
Yes, you're both right that there needs to be some arity information in the set. I'll update the PR. However, adding `Args` to the set as you suggest @zuiderkwast wouldn't be correct, I think. Consider the type below:

```
-type apa(A) :: A | apa({A}).
```

This type keeps changing the type of its argument so that we get an increasing level of nesting that we have never seen before and the normalization would go on forever.

Reply:
True. Unless A is unexpanded... I'm not sure. :-)